### PR TITLE
[pangram] exercise

### PR DESCRIPTION
**Pull request by leetwinski**

### CODE REVIEW

**Comment by leetwinski** [comment]
@exercism/clojure , added pangram exercise

File: exercises/pangram/src/example.clj [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+(def ^:private z-int (int \z))
+
+(def ^:private alphabet (set (range a-int (inc z-int))))
+
+(defn pangram? [input]
+  (= alphabet
+     (->> input
+          lower-case
+          (map int)
+          (filter #(<= a-int % z-int))
```
Comment (yurrriq) [inline exercises/pangram/src/example.clj]: Why not [`Character/isLowerCase`](http://docs.oracle.com/javase/8/docs/api/java/lang/Character.html#isLowerCase-char-) or regex here instead?

File: exercises/pangram/src/example.clj [hunk 2]
```diff
@@ -0,0 +1,15 @@
+(ns pangram
+  (:require [clojure.string :refer [lower-case]]))
+
+(def ^:private a-int (int \a))
+(def ^:private z-int (int \z))
+
+(def ^:private alphabet (set (range a-int (inc z-int))))
+
+(defn pangram? [input]
+  (= alphabet
```
Comment (yurrriq) [inline exercises/pangram/src/example.clj]: Why not fold the `(= alphabet)` into the `(->> ...)`?

File: exercises/pangram/src/example.clj [hunk 3]
```diff
@@ -0,0 +1,15 @@
+(ns pangram
+  (:require [clojure.string :refer [lower-case]]))
+
+(def ^:private a-int (int \a))
+(def ^:private z-int (int \z))
+
+(def ^:private alphabet (set (range a-int (inc z-int))))
```
Comment (yurrriq) [inline exercises/pangram/src/example.clj]: `a-int`, `z-int` and `alphabet` feel superfluous to me.

File: exercises/pangram/src/example.clj (hunk 1 above)
Comment (yurrriq) [inline exercises/pangram/src/example.clj]: ... or something like:

```clojure
(def char<=? (comp (complement pos?) compare))

(char<=? \a \z)
;; => true
(char<=? \a \a)
;; => true
(char<=? \z \a)
;; => false
```

File: exercises/pangram/test/pangram_test.clj [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  (is
+   (false?
+    (pangram? "7h3 qu1ck brown fox jumps ov3r 7h3 lazy dog"))))
+
+(deftest mixed-case-and-punctuation
+  (is (pangram? "\"Five quacking Zephyrs jolt my wax bed.\"")))
+
+(deftest non-ascii-characters
+  (is (pangram?
+       "Victor jagt zwölf Boxkämpfer quer über den großen Sylter Deich.")))
```
Comment (yurrriq) [inline exercises/pangram/test/pangram_test.clj]: Given `alphabet` and `#(<= a-int % z-int)` above, is the intended behaviour to ignore accented letters?

**Comment by yurrriq** [comment]
What are your thoughts on something like this?

```clojure
(ns pangram
  (:require [clojure.string :refer [lower-case]]
            [clojure.set    :refer [subset?]]))

(def ^:private alphabet (set "abcdefghijklmnopqrstuvwxyz"))

(defn pangram? [input]
  (-> (fn [acc chr]
        (cond-> acc
          (Character/isLetter chr) (conj (Character/toLowerCase chr))))
      (reduce #{} input)
      (->> (subset? alphabet))))
```

This way we only traverse `input` once and use `#'clojure.set/subset?`, rather than requiring equality.

File: exercises/pangram/test/pangram_test.clj (hunk 4 above)
Comment (leetwinski) [inline exercises/pangram/test/pangram_test.clj]: Yes, it should just consider the letters from the english alphabet. The rest is considered to be junk

File: exercises/pangram/src/example.clj (hunk 2 above)
Comment (leetwinski) [inline exercises/pangram/src/example.clj]: it should be. Thanks!

**Comment by leetwinski** [comment]
I don't know. As for me, i always find it `cond->` much harder to read. I would probably rewrite it to use `(= count 26)` as the correct solution indicator

**Comment by yurrriq** [comment]
I think folding over a set is the best solution, but have no strong feelings re: `cond->`. It could be a good use case for transducers too and it might be good to avoid the Java methods. `compare` works in cljs too fwiw.

**Comment by leetwinski** [comment]
@yurrriq , yet another variant: transducer + characters count.

**Comment by leetwinski** [comment]
Good point about clojurescript. Would rewrite `in-range?`

**Comment by yurrriq** [comment]
Nice. I like `char<=`. I still think we should be using sets and `#'clojure.set/subset?` rather than collecting a vector and checking its size.

Thoughts, @exercism/lisp?

**Comment by leetwinski** [comment]
I just feel uncomfortable about the double traversal (reduction + subset?) vs single pass with transducer and count. But I won't really insist if you say your variant is better, since for the values under test the performance doesn't matter i guess. Anyway, i can always post my answer as the solution straight to exercism :smile:

**Comment by yurrriq** [comment]
Fair point. `count` for vectors is O(1), right?

**Comment by leetwinski** [comment]
yes, since it is `counted`

**Comment by yurrriq** [comment]
My only other suggestion then would be to try to avoid `Character/toLowerCase` for portability. I'm happy to merge this now. What do you think?

**Comment by leetwinski** [comment]
The only thing i can think of, is to use `clojure.string/lower-case` on a char, but it seems to be an overhead. Maybe we should go with conditional compilation and rename example to `.cljc` ?

**Comment by yurrriq** [comment]
If you want to. Otherwise I think it's fine to leave it as it is, since we don't really mention portability anywhere. If anyone asks we can point to this discussion.

**Comment by leetwinski** [comment]
Ok then. I wouldn't insist on rewriting :smile:

**Comment by yurrriq** [comment]
Thanks for this!

**Pull request merged**